### PR TITLE
🔨

### PR DESCRIPTION
**Pull request by doggystylez**
price precision and logic needs to be verified with higher exp tokens, as well as ideal initial mint amounts
compounding of non vault rewards is not yet implemented
tests are passing which indicates the join/leave precision should be at least as much as the prior share implementation

## Summary by CodeRabbit

- **New Features**
  - Introduced `Metadata` struct and `Environment` enum for enhanced data handling.
  - Added new fields to `InstantiateMsg`, `ExecuteMsg`, and `Status` struct.

- **Bug Fixes**
  - Updated error messages for clarity and consistency.

- **Improvements**
  - Reorganized fields and messages for better readability and context.
  - Enhanced contract setup logic and balance calculations.
  - Updated dependencies for improved performance and compatibility.

- **Chores**
  - Updated Docker image version to `cosmwasm/optimizer:0.15.1`.
  - Updated package version in `Cargo.toml` and dependencies.

- - -
This change is [Reviewable](https://reviewable.io/reviews/banana-dao/banana-vaults/11)

### CODE REVIEW

File: src/contract.rs [hunk 1]
```diff
@@ -111,40 +121,64 @@ pub fn instantiate(
...
+    // create the vault token
+    let create_msg: CosmosMsg = MsgCreateDenom {
+        sender: env.contract.address.clone().into_string(),
+        subdenom: BASE_DENOM.to_string(),
+    }
+    .into();
+
+    // todo: refine initial mint logic
+    // provisional logic
+    let initial_mint = Uint128::new(DEC_18);
```
[the remark this answers is not in the document]
Comment (doggystylez) [reply src/contract.rs:143]: sure, we want to mint an initial supply to allow sufficient precision without risking overflow and also need to be conscious of share inflation attacks or similar exploits

File: src/contract.rs [hunk 2]
```diff
@@ -60,21 +69,23 @@ pub fn instantiate(
...
+        .pool(msg.pool_id)?
+        .pool
+    {
         Some(pool) => {
             if pool
                 .type_url
                 .ne(&"/osmosis.concentratedliquidity.v1beta1.Pool".to_string())
             {
-                return Err(ContractError::PoolIsNotCL {});
+                return Err(ContractError::PoolIsNotCL);
```
[the remark this answers is not in the document]
Comment (doggystylez) [reply src/contract.rs:88]: this is probably acceptable as it's only called on instantiate

File: src/contract.rs [hunk 3]
```diff
@@ -843,174 +826,164 @@ fn execute_force_exits(deps: DepsMut, env: &Env) -> Result<Response, ContractErr
...
 
-fn query_config(deps: Deps) -> StdResult<Config> {
-    let config = CONFIG.load(deps.storage)?;
-    Ok(config)
+fn query_locked_assets(deps: Deps, env: &Env) -> StdResult<Vec<Coin>> {
+    let address = env.contract.address.to_string();
+
+    let (asset0, asset1) = get_vault_balances(&deps, &address, true)?;
+
+    Ok(vec![asset0, asset1])
```
[the remark this answers is not in the document]
Comment (doggystylez) [reply src/contract.rs:874]: can you elaborate?

**Review by keyleu — CHANGES_REQUESTED** [review]
> Reviewed 2 of 7 files at r1, 1 of 5 files at r2, 2 of 4 files at r3, 2 of 2 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/banana-dao/banana-vaults/11)* status: all files reviewed, 14 unresolved discussions (waiting on @coderabbitai[bot] and @doggystylez)

___
*[`src/contract.rs` line 143 at r1](https://reviewable.io/reviews/banana-dao/banana-vaults/11#gh-1533427400:-NtW-IBz3F1MlV1mYDbH:bzd5hq) ([raw file](https://github.com/banana-dao/banana-vaults/blob/e0656d796ef421d72d9c8154aa076d3907364d38/src/contract.rs#L143)):*
<details><summary><i>Previously, coderabbitai[bot] wrote…</i></summary><blockquote>



Given the importance of ensuring both precision and security in the initial mint logic, it's crucial to consider the system's overall design and specific requirements for token supply and precision. Addressing overflow risks and preventing share inflation attacks will likely involve a combination of ensuring safe integer operations, validating minting conditions, and possibly implementing supply or mint rate caps.

To provide a more targeted recommendation, could you share more details about the system's architecture, the role of the token within the ecosystem, and any specific constraints or goals related to token supply and precision? This information will help in proposing a refined strategy that balances precision with security considerations.


</blockquote></details>

makes sense to have an initial mint, similar how liquidity pools mint an initial LP tokens when created, don't really see a problem with it.
___
*[`src/contract.rs` line 155 at r1](https://reviewable.io/reviews/banana-dao/banana-vaults/11#-NtW141d8syOP3CBhyNn:-NtW141d8syOP3CBhyNo:bntl47i) ([raw file](https://github.com/banana-dao/banana-vaults/blob/e0656d796ef421d72d9c8154aa076d3907364d38/src/contract.rs#L155)):*
> ```XML
>     .into();
> 
>     SUPPLY.save(deps.storage, &initial_mint)?;
> ```

You can query the supply directly from the token btw. Maybe it's better to query it because a user can burn his tokens permissionlessly (it's possible) and that money would be stuck forever in the vault.
___
*[`src/contract.rs` line 426 at r3](https://reviewable.io/reviews/banana-dao/banana-vaults/11#-NtXstqABwlAKaGjq4f6:-NtXstqABwlAKaGjq4f7:b-9sj3eu) ([raw file](https://github.com/banana-dao/banana-vaults/blob/87e55c71ddf7b34d2e08e3115acb3e5182de64ca/src/contract.rs#L426)):*
> ```XML
>     }
> 
>     // Check if vault is halted or cap has been reached
> ```

nit: fix comment
___
*[`src/contract.rs` line 434 at r3](https://reviewable.io/reviews/banana-dao/banana-vaults/11#-NtXso1S0AW4v0JLwtl3:-NtXso1S0AW4v0JLwtl4:b4m8ovr) ([raw file](https://github.com/banana-dao/banana-vaults/blob/87e55c71ddf7b34d2e08e3115acb3e5182de64ca/src/contract.rs#L434)):*
> ```XML
>     if CAP_REACHED.load(deps.storage)?
>         && WHITELISTED_DEPOSITORS
>             .may_load(deps.storage, info.sender.clone())?
> ```

nit: change to !WHITELISTED_DEPOSITORS.has(deps.storage, info.sender.clone())?
___
*[`src/contract.rs` line 443 at r3](https://reviewable.io/reviews/banana-dao/banana-vaults/11#-NtXsyML9msw7FOQHUG5:-NtXsyML9msw7FOQHUG6:b-zef2v9) ([raw file](https://github.com/banana-dao/banana-vaults/blob/87e55c71ddf7b34d2e08e3115acb3e5182de64ca/src/contract.rs#L443)):*
> ```XML
>     if ACCOUNTS_PENDING_BURN
>         .may_load(deps.storage, info.sender.clone())?
>         .is_some()
> ```

nit: change to ACCOUNTS_PENDING_BURN.has(deps.storage, info.sender.clone())?
___
*[`src/contract.rs` line 487 at r3](https://reviewable.io/reviews/banana-dao/banana-vaults/11#-NtXu_57E1tg-3olA57X:-NtXu_57E1tg-3olA57Y:bfzus6e) ([raw file](https://github.com/banana-dao/banana-vaults/blob/87e55c71ddf7b34d2e08e3115acb3e5182de64ca/src/contract.rs#L487)):*
> ```XML
>         .add_attribute("action", "banana_vault_deposit_for_mint")
>         .add_attribute("address", info.sender.to_string())
>         .add_attribute("amount", mint_assets[0].amount))
> ```

not sure if you are going to use these events for an indexer or smthing, but you're missing mint_assets[1] here.
___
*[`src/contract.rs` line 787 at r3](https://reviewable.io/reviews/banana-dao/banana-vaults/11#-NtY1hQm-FGk3ztRiNsX:-NtY1hQm-FGk3ztRiNsY:b35uxei) ([raw file](https://github.com/banana-dao/banana-vaults/blob/87e55c71ddf7b34d2e08e3115acb3e5182de64ca/src/contract.rs#L787)):*
> ```XML
>     let commission_rewards = COMMISSION_REWARDS.load(deps.storage)?;
> 
>     if commission_rewards.is_empty() {
> ```

This is not possible because you are always initializing with a vec![coin(0, config.asset0.denom), coin(0, config.asset1.denom)] so it will never
be empty.
___
*[`src/contract.rs` line 1128 at r3](https://reviewable.io/reviews/banana-dao/banana-vaults/11#-NtXzFENEnRC4ahohiG5:-NtXzFENEnRC4ahohiG6:b-awh7zp) ([raw file](https://github.com/banana-dao/banana-vaults/blob/87e55c71ddf7b34d2e08e3115acb3e5182de64ca/src/contract.rs#L1128)):*
> ```XML
>     for fund in funds {
>         if fund.denom == config.asset0.denom {
>             assets[0].amount = fund.amount;
> ```

should be += instead of =, since you can send the same denom twice with different amounts.
e.g. If you send [(1000, uosmo), (1,uosmo)], it would only consider 1 for minting and the rest would be lost. It's stupid but possible.
___
*[`src/contract.rs` line 1130 at r3](https://reviewable.io/reviews/banana-dao/banana-vaults/11#-NtXzb4V6xh6bVSXBoz7:-NtXzb4W80mHoOTnY_Tq:b2nreh6) ([raw file](https://github.com/banana-dao/banana-vaults/blob/87e55c71ddf7b34d2e08e3115acb3e5182de64ca/src/contract.rs#L1130)):*
> ```XML
>             assets[0].amount = fund.amount;
>         } else if fund.denom == config.asset1.denom {
>             assets[1].amount = fund.amount;
> ```

Same as above.
___
*[`src/contract.rs` line 1140 at r3](https://reviewable.io/reviews/banana-dao/banana-vaults/11#-NtY-cl26OPa7o9oZaUg:-NtY-cl26OPa7o9oZaUh:begapra) ([raw file](https://github.com/banana-dao/banana-vaults/blob/87e55c71ddf7b34d2e08e3115acb3e5182de64ca/src/contract.rs#L1140)):*
> ```XML
> 
> fn verify_burn_funds(storage: &dyn Storage, funds: &[Coin]) -> Result<(), ContractError> {
>     if funds.is_empty() {
> ```

nit: There is a one_coin(&funds) function in the cw-utils crate that does the logic you do here and below.
___
*[`src/contract.rs` line 1607 at r3](https://reviewable.io/reviews/banana-dao/banana-vaults/11#-NtY5xWjEqz4aW4f8SEq:-NtY5xWjEqz4aW4f8SEr:b9jtw4s) ([raw file](https://github.com/banana-dao/banana-vaults/blob/87e55c71ddf7b34d2e08e3115acb3e5182de64ca/src/contract.rs#L1607)):*
> ```XML
>     SUPPLY.save(deps.storage, &(supply.checked_add(total_minted)?))?;
> 
>     ASSETS_PENDING_MINT.save(deps.storage, &pending_assets)?;
> ```

This will always be [(0,asset1.denom),[(0,asset2.denom)], correct?
___
*[`src/contract.rs` line 1701 at r4](https://reviewable.io/reviews/banana-dao/banana-vaults/11#-NtY8U-32s8lZAnb8DKi:-NtY8U-32s8lZAnb8DKj:by06qbs) ([raw file](https://github.com/banana-dao/banana-vaults/blob/bfb8563951f755432dc41804c488609b35cf9f44/src/contract.rs#L1701)):*
> ```XML
>     {
>         return Err(ContractError::CantProcessBurn);
>     }
> ```

I guess this is just a safety mechanism and should never happen, correct?

**Review by doggystylez — COMMENTED** [review]
> Reviewed 1 of 2 files at r4.
> *[Reviewable](https://reviewable.io/reviews/banana-dao/banana-vaults/11)* status: 6 of 7 files reviewed, 11 unresolved discussions (waiting on @coderabbitai[bot] and @keyleu)

___
*[`src/contract.rs` line 211 at r2](https://reviewable.io/reviews/banana-dao/banana-vaults/11#gh-1534109065:-NtYVh1n-H0gVRaYIBUL:b-896fix) ([raw file](https://github.com/banana-dao/banana-vaults/blob/5bf05f1c59a0ecbdf4534ca1b0bd7b02b991b10c/src/contract.rs#L211)):*
<details><summary><i>Previously, coderabbitai[bot] wrote…</i></summary><blockquote>

The `execute` function's switch-like behavior on `ExecuteMsg` variants is clear and well-structured. However, considering the function's length and complexity, it might be beneficial to refactor each case into separate functions or modules. This would improve readability and maintainability, making it easier to understand and modify individual behaviors without affecting others.


</blockquote></details>

Done.
___
*[`src/contract.rs` line 874 at r2](https://reviewable.io/reviews/banana-dao/banana-vaults/11#gh-1534109070:-NtYVucUCpa857gVu6bE:b-896fix) ([raw file](https://github.com/banana-dao/banana-vaults/blob/5bf05f1c59a0ecbdf4534ca1b0bd7b02b991b10c/src/contract.rs#L874)):*
<details><summary><i>Previously, coderabbitai[bot] wrote…</i></summary><blockquote>



The suggestion to abstract the logic for retrieving vault balances into a reusable function aims to enhance code maintainability and reduce duplication. In the provided code, the `query_locked_assets` function retrieves the locked assets by calling `get_vault_balances` with specific parameters. This pattern of retrieving vault balances is a prime candidate for abstraction, especially if similar logic is used in multiple places within the contract.

By abstracting this logic into a reusable function, you can achieve several benefits:
- **Reduce Code Duplication**: If the logic for retrieving vault balances is repeated in multiple places, abstracting it into a single function reduces the amount of duplicated code. This makes the codebase smaller and easier to understand.
- **Improve Maintainability**: When the logic for retrieving vault balances needs to be updated or fixed, having it in one place means you only need to make changes in one location. This reduces the risk of inconsistent behavior across different parts of the contract.
- **Enhance Readability**: A well-named function can make the code more readable by clearly indicating what the function does. For example, a function named `getVaultBalances` immediately conveys its purpose.

A potential implementation for the abstraction could look like this:
```rust
fn get_vault_balances(deps: Deps, include_position: bool) -> StdResult<(Coin, Coin)> {
    // Implementation logic here...
}
```
This function encapsulates the logic for retrieving vault balances and can be called from various parts of the contract with the appropriate parameters. This approach ensures that the logic is centralized, making the contract easier to maintain and modify in the future.


</blockquote></details>

Done.
___
*[`src/contract.rs` line 487 at r3](https://reviewable.io/reviews/banana-dao/banana-vaults/11#-NtXu_57E1tg-3olA57X:-NtYRB9p3_EsjfqJKjM2:bvblu7f) ([raw file](https://github.com/banana-dao/banana-vaults/blob/87e55c71ddf7b34d2e08e3115acb3e5182de64ca/src/contract.rs#L487)):*
<details><summary><i>Previously, keyleu (Keyne) wrote…</i></summary><blockquote>

not sure if you are going to use these events for an indexer or smthing, but you're missing mint_assets[1] here.
</blockquote></details>

these can be removed 
___
*[`src/contract.rs` line 787 at r3](https://reviewable.io/reviews/banana-dao/banana-vaults/11#-NtY1hQm-FGk3ztRiNsX:-NtYRgUrCQ7iW1nmQWan:b-4brhjz) ([raw file](https://github.com/banana-dao/banana-vaults/blob/87e55c71ddf7b34d2e08e3115acb3e5182de64ca/src/contract.rs#L787)):*
<details><summary><i>Previously, keyleu (Keyne) wrote…</i></summary><blockquote>

This is not possible because you are always initializing with a vec![coin(0, config.asset0.denom), coin(0, config.asset1.denom)] so it will never
be empty.
</blockquote></details>

replaced with a check for is_zero()
___
*[`src/contract.rs` line 1128 at r3](https://reviewable.io/reviews/banana-dao/banana-vaults/11#-NtXzFENEnRC4ahohiG5:-NtYUXPS-83685EV8R4R:b-kr97eq) ([raw file](https://github.com/banana-dao/banana-vaults/blob/87e55c71ddf7b34d2e08e3115acb3e5182de64ca/src/contract.rs#L1128)):*
<details><summary><i>Previously, keyleu (Keyne) wrote…</i></summary><blockquote>

should be += instead of =, since you can send the same denom twice with different amounts.
e.g. If you send [(1000, uosmo), (1,uosmo)], it would only consider 1 for minting and the rest would be lost. It's stupid but possible.
</blockquote></details>

I think this shouldn't happen as `verify\_mint\_funds()` is called with info.funds (which should always be valid coins) in the execute messages. it would be possible to do it in the query.
___
*[`src/contract.rs` line 1140 at r3](https://reviewable.io/reviews/banana-dao/banana-vaults/11#-NtY-cl26OPa7o9oZaUg:-NtYTO9993y2qXOCiFH9:brux4q3) ([raw file](https://github.com/banana-dao/banana-vaults/blob/87e55c71ddf7b34d2e08e3115acb3e5182de64ca/src/contract.rs#L1140)):*
<details><summary><i>Previously, keyleu (Keyne) wrote…</i></summary><blockquote>

nit: There is a one_coin(&funds) function in the cw-utils crate that does the logic you do here and below.
</blockquote></details>

👀 it also has must_pay() which require the exact denom also. however they require the message itself not the funds directly
___
*[`src/contract.rs` line 1607 at r3](https://reviewable.io/reviews/banana-dao/banana-vaults/11#-NtY5xWjEqz4aW4f8SEq:-NtYSoMR-Ix0IedqJcLb:b-5pmre0) ([raw file](https://github.com/banana-dao/banana-vaults/blob/87e55c71ddf7b34d2e08e3115acb3e5182de64ca/src/contract.rs#L1607)):*
<details><summary><i>Previously, keyleu (Keyne) wrote…</i></summary><blockquote>

This will always be [(0,asset1.denom),[(0,asset2.denom)], correct?
</blockquote></details>

because of the optional slippage check above we can refuse to process some assets and they will remain there:

src/contract.rs:1581
___
*[`src/contract.rs` line 1701 at r4](https://reviewable.io/reviews/banana-dao/banana-vaults/11#-NtY8U-32s8lZAnb8DKi:-NtYS1xqCBpSu_9AB-Zn:bz26z9m) ([raw file](https://github.com/banana-dao/banana-vaults/blob/bfb8563951f755432dc41804c488609b35cf9f44/src/contract.rs#L1701)):*
<details><summary><i>Previously, keyleu (Keyne) wrote…</i></summary><blockquote>

I guess this is just a safety mechanism and should never happen, correct?
</blockquote></details>

we are now allowing burns when the position is open, so in some cases we will have enough liquid to repay them without withdrawing from the position

**Review by doggystylez — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/banana-dao/banana-vaults/11)* status: 6 of 7 files reviewed, 11 unresolved discussions (waiting on @coderabbitai[bot] and @keyleu)

___
*[`src/contract.rs` line 787 at r3](https://reviewable.io/reviews/banana-dao/banana-vaults/11#-NtY1hQm-FGk3ztRiNsX:-NtYZ7s-0taliN2gMzrM:bpp5bff) ([raw file](https://github.com/banana-dao/banana-vaults/blob/87e55c71ddf7b34d2e08e3115acb3e5182de64ca/src/contract.rs#L787)):*
<details><summary><i>Previously, doggystylez wrote…</i></summary><blockquote>

replaced with a check for is_zero()
</blockquote></details>

nvm, instead we'll remove zero coins first as they shouldnt be sent

**Review by doggystylez — COMMENTED** [review]
> Reviewed all commit messages.
> Dismissed @coderabbitai[bot] from 2 discussions.
> *[Reviewable](https://reviewable.io/reviews/banana-dao/banana-vaults/11)* status: 6 of 7 files reviewed, 9 unresolved discussions (waiting on @coderabbitai[bot] and @keyleu)

___
*[`src/contract.rs` line 88 at r2](https://reviewable.io/reviews/banana-dao/banana-vaults/11#gh-1534109061:-NtYZY8aCx-0LmeWZdR1:b-896fix) ([raw file](https://github.com/banana-dao/banana-vaults/blob/5bf05f1c59a0ecbdf4534ca1b0bd7b02b991b10c/src/contract.rs#L88)):*
<details><summary><i>Previously, coderabbitai[bot] wrote…</i></summary><blockquote>



@doggystylez, I understand your point about the instantiation context potentially mitigating the risks associated with using `unwrap()`. It's true that the controlled nature of instantiation might reduce the likelihood of encountering unexpected conditions that could lead to panics. However, it's generally advisable to adhere to best practices by avoiding `unwrap()` when possible, even in controlled contexts. This approach enhances the robustness and error tolerance of the contract. If the operation preceding `unwrap()` is guaranteed not to fail under the conditions present at instantiation, it might be helpful to either handle the error explicitly or document the rationale behind this assumption for clarity and maintainability.


</blockquote></details>

Done.
___
*[`src/contract.rs` line 88 at r4](https://reviewable.io/reviews/banana-dao/banana-vaults/11#gh-1534467382:-NtYZTT12KvXHowYACqv:b-896fix) ([raw file](https://github.com/banana-dao/banana-vaults/blob/bfb8563951f755432dc41804c488609b35cf9f44/src/contract.rs#L88)):*
<details><summary><i>Previously, coderabbitai[bot] wrote…</i></summary><blockquote>

Using `unwrap()` in a production contract, especially without prior validation, can lead to panics and unexpected contract termination. Consider handling the error case explicitly or ensuring that the operation cannot fail before calling `unwrap()`.


</blockquote></details>

Done.
___
*[`src/contract.rs` line 143 at r4](https://reviewable.io/reviews/banana-dao/banana-vaults/11#gh-1534467386:-NtYZbSN53nfubVqEmS5:b-896fix) ([raw file](https://github.com/banana-dao/banana-vaults/blob/bfb8563951f755432dc41804c488609b35cf9f44/src/contract.rs#L143)):*
<details><summary><i>Previously, coderabbitai[bot] wrote…</i></summary><blockquote>

The comment "todo: refine initial mint logic" suggests that the initial mint logic is provisional and may need further refinement. It's important to address this TODO to ensure that the initial minting process aligns with the intended logic and security considerations.

Would you like assistance in refining the initial mint logic?


</blockquote></details>

Done.

**Review by keyleu — APPROVED** [review]
> Reviewed 1 of 1 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/banana-dao/banana-vaults/11)* status: all files reviewed, 9 unresolved discussions (waiting on @coderabbitai[bot] and @doggystylez)

___
*[`src/contract.rs` line 1128 at r3](https://reviewable.io/reviews/banana-dao/banana-vaults/11#-NtXzFENEnRC4ahohiG5:-NtYcCpP9qZLX8tlOLPU:bw5d40) ([raw file](https://github.com/banana-dao/banana-vaults/blob/87e55c71ddf7b34d2e08e3115acb3e5182de64ca/src/contract.rs#L1128)):*
<details><summary><i>Previously, doggystylez wrote…</i></summary><blockquote>

I think this shouldn't happen as `verify\_mint\_funds()` is called with info.funds (which should always be valid coins) in the execute messages. it would be possible to do it in the query.
</blockquote></details>

correct, thought you could send duplicates

**Review by doggystylez — COMMENTED** [review]
> Dismissed @coderabbitai[bot] from 6 discussions.
> *[Reviewable](https://reviewable.io/reviews/banana-dao/banana-vaults/11)* status: all files reviewed, 3 unresolved discussions (waiting on @coderabbitai[bot])

**Pull request merged**